File: interviewBit-golang/src/backtracking/problems/permutations.go
package backtracking

/*
	Problem : https://www.interviewbit.com/problems/permutations/
*/

func solvePermutations(A []int, B *[][]int, idx, n int) {
	if idx == n {
		C := make([]int, n)
		copy(C, A)
		*B = append(*B, C)
		return
	}

	for i := idx; i < n; i++ {
		A[i], A[idx] = A[idx], A[i]
		solvePermutations(A, B, idx+1, n)
		A[i], A[idx] = A[idx], A[i]
	}
}

// T(n) : O() ; S(n) : O()
func Permutations(A []int) [][]int {

	n := len(A)
	res := make([][]int, 0)
	if n == 0 {
		return res
	}

	solvePermutations(A, &res, 0, n)
	return res
}

File: interviewBit-cpp/src/backtracking/permutations.cpp
#include <bits/stdc++.h>

using namespace std;

// https://www.interviewbit.com/problems/permutations/
class Solution {
	vector<vector<int> > permute(vector<int> &A);
};

void backtrackSolvePermute(size_t idx, vector<int> &A, vector<vector<int> > &allPermutations) {
	
	if (idx >= A.size()) {
		allPermutations.push_back(A);
		return ;
	}
	
	for (size_t i = idx; i < A.size(); i++) {
		swap(A[idx], A[i]);
		backtrackSolvePermute(idx + 1, A, allPermutations);
		swap(A[idx], A[i]);
	}
}

// Approach 1
// T(n) : O(n*n!) ; S(n) : O(n)
vector<vector<int> > solvePermute(vector<int> &A) {
	
	size_t n = A.size();
	vector<vector<int> > allPermutations;
	if (n == 0) return allPermutations;
	
	size_t idx = 0;
	backtrackSolvePermute(idx, A, allPermutations);
	return allPermutations; 
}

void backtrackSolvePermute1(vector<int> &curr, vector<int> &A, vector<vector<int> > &allPermutations) {
	
	if (curr.size() == A.size()) {
		allPermutations.push_back(curr);
		return ;
	}
	
	for (auto &a : A) {
		if (find(curr.begin(), curr.end(), a) == curr.end()) {
			curr.push_back(a);
			backtrackSolvePermute1(curr, A, allPermutations);
			curr.pop_back();
		}
	}
}

// Approach 2
// T(n) : O(n*n!) ; S(n) : O(n)
vector<vector<int> > solvePermute1(vector<int> &A) {
	
	size_t n = A.size();
	vector<vector<int > > allPermutations;
	if (n == 0) return allPermutations;
	
	vector<int> temp = {};
	backtrackSolvePermute1(temp, A, allPermutations);
	return allPermutations;
}

vector<vector<int> > Solution::permute(vector<int> &A) {
	return solvePermute1(A);
}

// Driver Code for testing
int main() {

    return 0;
}
